### PR TITLE
[CI] Enable packaging with Pipeline 2.0

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Proposal to run the packaging in parallel on a PR basis.

See the follow ups section regarding what are the different areas that could be done in the future.

## Why is it important?

Split the packaging per beat in order to validate them on a PR basis.

## Tasks

* [x] Test
* [x] E2E disabled at the moment but already implemented
* [x] publish packages and docker images

## Follow ups
* Enable e2e
* ARM support if possible
* MAC support if possible
* Deprecate the existing packaging pipeline in favour of this approach

## Issues

Requires https://github.com/elastic/beats/pull/23960
Caused https://github.com/elastic/e2e-testing/pull/705

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -259,6 +259,61 @@ def k8sTest(Map args = [:]) {
...
+    try {
+      if(isInstalled(tool: 'docker', flag: '--version')) {
+        dockerLogin(secret: "${DOCKER_ELASTIC_SECRET}", registry: "${DOCKER_REGISTRY}")
+      }
+      retry(3){
+        sh script: """.ci/scripts/install-test-dependencies.sh "${suite}" """, label: "Install test dependencies for ${suite}:${tags}"
+      }
+      filebeat(output: "docker_logs_${suite}_${tags}.log", workdir: "${env.WORKSPACE}"){
+        // TBC with the suite to be used
+        sh script: """SUITE=metricbeat DEVELOPER_MODE=false TIMEOUT_FACTOR=3 LOG_LEVEL=TRACE make -C e2e functional-test""", label: "Run functional tests for ${suite}:${tags}"
```
Comment (mdelapenya) [inline Jenkinsfile:306]:
```suggestion
        sh script: """SUITE="${suite}" TAGS="${tags}" DEVELOPER_MODE=false TIMEOUT_FACTOR=3 LOG_LEVEL=TRACE make -C e2e functional-test""", label: "Run functional tests for ${suite}:${tags}"
```

Comment (v1v) [reply Jenkinsfile:306]: We need to discuss about this, its' only for testing purposes now, but need to find a way to simplify the commands with a simplistic make goal or something like that.

File: Jenkinsfile [hunk 2]
```diff
@@ -259,6 +262,61 @@ def k8sTest(Map args = [:]) {
...
+                'linux/arm64',
+                'linux/armv7',
+                // The platforms above are disabled temporarly as crossbuild images are
+                // not available. See: https://github.com/elastic/golang-crossbuild/issues/71
+                //'linux/ppc64le',
+                //'linux/mips64',
+                //'linux/s390x',
+                'windows/amd64',
+                'windows/386',
+                (params.macos ? '' : 'darwin/amd64'),
```
Comment (v1v) [inline Jenkinsfile:281]: Let's see what we do with this

File: x-pack/heartbeat/Jenkinsfile.yml [hunk 3]
```diff
@@ -0,0 +1,18 @@
+when:
```
Comment (v1v) [inline x-pack/heartbeat/Jenkinsfile.yml:1]: This is a new step, though there is no build in place yet but the packaging itself.

File: Jenkinsfile [hunk 4]
```diff
@@ -66,6 +66,9 @@ pipeline {
     }
     stage('Lint'){
       options { skipDefaultCheckout() }
+      when {
+        expression { return false }
+      }
```
Comment (v1v) [inline Jenkinsfile:71]: To speed up the testing

File: Jenkinsfile [hunk 5]
```diff
@@ -259,6 +262,61 @@ def k8sTest(Map args = [:]) {
...
+* This method runs the end 2 end testing in the same worker where the packages have been
+* generated, this should help to speed up the things
+*/
+def e2e(Map args = [:]) {
+  def suite = args.get('suite', '')
+  def tags = args.get('tags', '')
+  def stackVersion = args.get('stackVersion', '8.0.0-SNAPSHOT')  // TBC with the version defined somewhere...
+  dir("${env.WORKSPACE}/src/github.com/elastic/e2e-testing") {
+    // TBC with the target branch if running on a PR basis.
+    git(branch: 'feature/metricbeat-goal', credentialsId: '2a9602aa-ab9f-4e52-baf3-b71ca88469c7-UserAndToken', url: 'https://github.com/elastic/e2e-testing.git')
```
Comment (v1v) [inline Jenkinsfile:300]: TBC

File: metricbeat/Jenkinsfile.yml [hunk 6]
```diff
@@ -75,3 +75,8 @@ stages:
         mage: "mage build unitTest"
         platforms:             ## override default labels in this specific stage.
             - "windows-7-32-bit"
+    packaging-linux:
+        packaging-linux: "mage package"
+        e2e:
+            enabled: true
+            entrypoint: 'metricbeat-test.sh'
```
Comment (v1v) [inline metricbeat/Jenkinsfile.yml:82]: This is already implemented with:
- https://github.com/elastic/e2e-testing/pull/705

File: x-pack/elastic-agent/Jenkinsfile.yml [hunk 7]
```diff
@@ -85,3 +85,7 @@ stages:
     #            - "windows-7-32"
     #        branches: true     ## for all the branches
     #        tags: true         ## for all the tags
+    packaging-linux:
+        packaging-linux: "mage package"
+        e2e:
+            enabled: false
```
Comment (EricDavisX) [inline x-pack/elastic-agent/Jenkinsfile.yml:91]: is this where we'd set it to true to run the e2e-testings for Agent?  @mdelapenya @v1v not sure what else we need to do here, but along with all the Beats, Agent urgently needs this to help keep tests passing and quality quantified.

Comment (v1v) [reply x-pack/elastic-agent/Jenkinsfile.yml:91]: @EricDavisX , this is an improvement for running the packaging with e2e within the main pipeline. So far, I'm working on iterations, so testing one of the beats at the time. 

This improvement and the existing one are complementaries, and I won't change the existing workflow, until we feel safe to do so.

The existing workflow runs as below:

* on branches, it runs the main pipeline (build/lint/testing), then the packaging pipeline and finally the e2e pipeline.
* on pr basis, it runs the main pipeline, but the packaging and e2e pipelines are not triggered by default, it requires a GitHub comment to enable it.

This proposal want to improve the automation on a PR basis and skip the GitHub comments, since not every contributor uses it and worth to run the e2e on a PR basis to catch errors before any merge.

Please let me know if I need to clarify this further

**Comment by urso** [comment]
Does this change mean the `pr-merge` will only be green if packaging and E2E tests did succeed? Especially E2E tests are quite red recently and flakyness in Beats tests has increased again.

With this change, will we run all E2E tests? E.g. if I modify filebeat, will metricbeat packaing or auditbeat E2E test be able to mark my PR as failed?

Regarding running packaging automatically. I would expect that packaging jobs are run when any magefile, Makfile, dev-tools or scripts directory is touched in a PR.

**Comment by v1v** [comment]
> Does this change mean the pr-merge will only be green if packaging and E2E tests did succeed? Especially E2E tests are quite red recently and flakyness in Beats tests has increased again.

Exactly, if the concern is regarding the E2E flakiness, I think there were a couple of issues here:
- some regressions that were not detected since the master branch got broken. (See later on an example of this particular scenario)
- packages were not generated for `x-pack` for the last 3 days. 

This particular approach will help to detect earlier any broken changes before any merges.

cc @mdelapenya as he can clarify the above.

> With this change, will we run all E2E tests? E.g. if I modify filebeat, will metricbeat packaing or auditbeat E2E test be able to mark my PR as failed?

No, E2E will only run for those beats that got E2E to run and only if the changes are related to that particular beat. In other words, if you modify `filebeat` then only the packaging for filebeat will be executed, and if there are E2E for filebeat then it will be triggered afterwards, if no `E2E` then nothing will happen but the packaging.

> Regarding running packaging automatically. I would expect that packaging jobs are run when any magefile, Makfile, dev-tools or scripts directory is touched in a PR.

Both systems will coexist and eventually we will move to this approach. The more couple the build/test/packaging/e2e stages are the sooner we can find regressions/bugs on a PR basis.

The current approach for every commit in the `master` branch, or any, follow the below steps:
* Build/Test/IngTest all the beats.
  * If the above passed then, it runs the `Packaging and E2E` stage for each beat.
  * If it does not pass, then no `Packaging and E2E` will be executed

If no packaging and E2E are executed after a few commits, then it means the E2E might not detect regressions and the git bisect between the last successful `Packaging and E2E` and the latest commit might not be easy to do so.

In order to help with the analysis, we want to shift left all the validations that run by default in any branch. And ensure those validations are synchronous, and move away from the perception that merging on a broken branch does not produce slowness to everyone.

In a nutshell, this process does nothing but validating the build/test/cross-compile/ing-testing/packaging/e2e in parallel for the specific beat. Therefore it's similar what we got already in place but with the addition of the packaging, the e2e will only happen if there is something configured for that particular beat.

**Comment by urso** [comment]
> The current approach for every commit in the master branch, or any, follow the below steps:
>
>    Build/Test/IngTest all the beats.
>    - If the above passed then, it runs the Packaging and E2E stage for each beat.
>    - If it does not pass, then no Packaging and E2E will be executed
>
> If no packaging and E2E are executed after a few commits, then it means the E2E might not detect regressions and the git bisect between the last successful Packaging and E2E and the latest commit might not be easy to do so.

+1 on running packaging, E2E and tests in parallel. When testing branches, no tests should be excluded.

> - some regressions that were not detected since the master branch got broken. (See later on an example of this particular scenario)
> - packages were not generated for x-pack for the last 3 days.
>
> This particular approach will help to detect earlier any broken changes before any merges.

Again +1 for always running all CI on master and release/development branches.

This also shows shortcomming in alerting. We should be able to build from master and 7.x at any time. If builds go red, the sirens should turn on. No packages for 3 days without anyone investigating for 3 workdays is unacceptable. We introduced a Build Manager Rotation last week having Beats developers help in analyzing and fixing tests. Are we missing discoverability/alerting?


> No, E2E will only run for those beats that got E2E to run and only if the changes are related to that particular beat. In other words, if you modify filebeat then only the packaging for filebeat will be executed, and if there are E2E for filebeat then it will be triggered afterwards, if no E2E then nothing will happen but the packaging.

+1 on testing only based on related changes. If I modify filebeat, will x-pack/filebeat packaging and E2E tests be run as well?
Changing metricbeat internals might also require auditbeat testsuite to be run.


In general there are serious issues in the process that need to be addressed. Not just during PRs but regular alerting as well, and us getting active. I see how the approach can help, but given that CI is not as stable as it should be yet, I worry about running everything on each PR potentially affecting the developer experience negatively, as we will need to run more things in parallel and increase the amount of sometimes unrelated tests that must pass.

File: Jenkinsfile [hunk 8]
```diff
@@ -74,15 +75,15 @@ pipeline {
           withBeatsEnv(archive: false, id: "lint") {
             dumpVariables()
             setEnvVar('VERSION', sh(label: 'Get beat version', script: 'make get-version', returnStdout: true)?.trim())
-            whenTrue(env.ONLY_DOCS == 'true') {
+            /*whenTrue(env.ONLY_DOCS == 'true') {
```
Comment (v1v) [inline Jenkinsfile:78]:
```suggestion
            whenTrue(env.ONLY_DOCS == 'true') {
```

File: Jenkinsfile [hunk 9]
```diff
@@ -74,15 +75,15 @@ pipeline {
...
               cmd(label: "make check", script: "make check")
             }
             whenTrue(env.ONLY_DOCS == 'false') {
               cmd(label: "make check-python", script: "make check-python")
               cmd(label: "make check-go", script: "make check-go")
               cmd(label: "make notice", script: "make notice")
               cmd(label: "Check for changes", script: "make check-no-changes")
-            }
+            }*/
```
Comment (v1v) [inline Jenkinsfile:86]:
```suggestion
            }
```

**Review by urso — APPROVED** [review]

**Review by andresrc — APPROVED** [review]
